### PR TITLE
Engine Scripts: links and main start page on engine start

### PR DESCRIPTION
**Pull request by RomanYarik**
@AviVahl, this is not ready yet.

The react part will be refactored and obviously styled :)

### CODE REVIEW

File: packages/scripts/src/node-environments-magager.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                await this.runNodeEnvironmentFunction({ featureName: featureName!, configName })
+            );
+        }
+    }
+
+    public async closeFeature({ featureName }: IRunOptions) {
+        if (!featureName) {
+            throw new Error('feature name was not provided');
+        }
+        if (!this.runningEnvironments.has(featureName!)) {
```
Comment (AviVahl) [inline packages/scripts/src/node-environments-magager.ts]: get than check res

File: packages/scripts/src/node-environments-magager.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            throw new Error(`there are no node environments running for ${featureName}`);
+        }
+        const runningFeatureEnvironments = this.runningEnvironments.get(featureName!)!;
+        await runningFeatureEnvironments.close();
+        this.runningEnvironments.delete(featureName!);
+    }
+
+    public getRunningFeatures({ featureName }: IRunOptions = {}) {
```
Comment (AviVahl) [inline packages/scripts/src/node-environments-magager.ts]: refactor per offline discussion

File: packages/scripts/src/node-environments-magager.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            const { configName, featureName }: IRunOptions = req.query;
+            try {
+                await this.runFeature({ configName, featureName });
+                res.json({
+                    result: 'success'
+                });
+            } catch (error) {
+                res.status(404).json({
+                    result: 'error',
+                    error: error.message
```
Comment (AviVahl) [inline packages/scripts/src/node-environments-magager.ts]: error && error.message

File: packages/scripts/test/node-environments-manager.unit.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        const app = new Application(nodeEnvironmentFixturePath);
+        const { close, nodeEnvironmentManager } = await app.start();
+
+        await nodeEnvironmentManager.runFeature(runFeatureOptions);
+
+        disposables.add(() => nodeEnvironmentManager.closeFeature(runFeatureOptions));
+        disposables.add(() => close());
+
+        const allOpenEnvironments = await nodeEnvironmentManager.getRunningFeatures();
+        expect(allOpenEnvironments).to.be.not.an('undefined');
```
Comment (AviVahl) [inline packages/scripts/test/node-environments-manager.unit.ts]: remove line

File: packages/scripts/test/node-environments-manager.unit.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        expect(await nodeEnvironmentManager.getRunningFeatures()).to.contain(runFeatureOptions.featureName);
+
+        disposables.add(() => nodeEnvironmentManager.closeFeature(runFeatureOptions));
+        disposables.add(() => close());
+    });
+
+    it('fails to launch if wrong config name or feature name are provided', async () => {
+        const app = new Application(nodeEnvironmentFixturePath);
+        const { close, nodeEnvironmentManager } = await app.start();
+        expect(nodeEnvironmentManager.runFeature({})).to.eventually.throw('feature name was not provided');
```
Comment (AviVahl) [inline packages/scripts/test/node-environments-manager.unit.ts]: await

File: packages/scripts/test/node-environments-manager.unit.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        expect(nodeEnvironmentManager.runFeature({ featureName: 'test' })).to.eventually.throw(
+            'config name was not provided'
+        );
+        expect(nodeEnvironmentManager.runFeature({ featureName: 'test' })).to.eventually.throw();
+        disposables.add(() => close());
```
Comment (AviVahl) [inline packages/scripts/test/node-environments-manager.unit.ts]: move this line right after the await that returned close().
do for all

File: packages/scripts/src/engine-dashboard/components/dashboard.tsx [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import { FeaturesList } from './features-list';
+
+export const Dashboard: React.FunctionComponent = () => {
+    return (
+        <div>
```
Comment (AviVahl) [inline packages/scripts/src/engine-dashboard/components/dashboard.tsx]: `<></>`

File: packages/scripts/src/create-webpack-configs.ts [hunk 8]
```diff
@@ -32,7 +32,17 @@ export function createWebpackConfigs(options: ICreateWebpackConfigsOptions): web
...
+                plugins: [
+                    new HtmlWebpackPlugin({
+                        filename: `index.html`,
+                        chunks: ['index']
+                    }),
+                    new VirtualModulesPlugin(virtualModules),
+                    new StylableWebpackPlugin()
+                ],
+                entry: {
+                    index: fs.join(__dirname, 'engine-dashboard', 'index.tsx')
```
Comment (AviVahl) [inline packages/scripts/src/create-webpack-configs.ts]: `require.resolve('./engi ...')`

File: packages/scripts/src/engine-dashboard/components/features-list.tsx [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+const fetchAllPossibleFeatures = async () => (await fetch('possible-entities')).json();
+const fetchRunningNodeEnvironments = async () => (await fetch('node-env')).json();
+
+export const FeaturesList: React.FunctionComponent = () => {
+    const [featuresConfigsList, setFeaturesConfigsList] = useState<PossibleFeaturesAndConfigs>({
+        features: [],
+        configs: []
+    });
+
+    const [runningNodeEnvironments, setRunningNodeEnvironments] = useState<string[]>([]);
```
Comment (AviVahl) [inline packages/scripts/src/engine-dashboard/components/features-list.tsx]: use Set<string>

File: packages/scripts/src/engine-dashboard/components/features-list.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    const onServerEnvironmentStatusChange = (
+        serverResponse: ServerResponse,
+        featureName: string,
+    ) => {
+        if (isSuccessResponse(serverResponse)) {
+            if (!runningNodeEnvironments.includes(featureName)) {
```
Comment (AviVahl) [inline packages/scripts/src/engine-dashboard/components/features-list.tsx]: get envs from server

File: packages/scripts/src/engine-dashboard/components/button.tsx [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    onClick: (response: ServerResponse, featureName: string, configName: string) => void;
+}
+
+export const ServerEnvironmentButton: React.FunctionComponent<ButtonProps> = ({
+    isServerActive,
+    featureName,
+    configName,
+    onClick
+}) => {
+    const changeServerState = async () =>
```
Comment (AviVahl) [inline packages/scripts/src/engine-dashboard/components/button.tsx]: rename. "node env"

File: packages/scripts/src/engine-dashboard/components/button.tsx [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    const onButtonClick = async () => {
+        const response = await changeServerState();
+        if (isServerResponseMessage(response)) {
+            onClick(response, featureName, configName);
+        } else {
+            throw new Error(`Unexpected response from server: ${response}`);
+        }
+    };
+
+    const buttonText = `${!isServerActive ? 'Start' : 'Close'} server environment(s)`;
```
Comment (AviVahl) [inline packages/scripts/src/engine-dashboard/components/button.tsx]: trinary in var text in jsx

File: packages/scripts/src/application.ts [hunk 13]
```diff
@@ -207,7 +221,23 @@ export class Application {
...
+
+    private getConfigNamesForRunningFeatures(
+        features: Map<string, IFeatureDefinition>,
+        configurations: Map<string, string>
+    ) {
+        const packageToConfigurationMapping: { features: string[]; configs: string[] } = {
+            configs: Array.from(configurations.keys()),
+            features: []
+        };
+        for (const [, { scopedName, isRoot }] of features) {
```
Comment (AviVahl) [inline packages/scripts/src/application.ts]: refactor to one liner
`Array.from(features.values()).filter({}) ...`

File: packages/scripts/package.json [hunk 14]
```diff
@@ -32,6 +32,9 @@
     "express": "^4.17.1",
     "glob": "^7.1.4",
     "html-webpack-plugin": "^4.0.0-beta.8",
+    "react": "^16.8.6",
+    "react-dom": "^16.8.6",
+    "react-table": "^6.10.0",
```
Comment (AviVahl) [inline packages/scripts/package.json]: add prop-types. remove from root devDeps

File: packages/scripts/src/application.ts [hunk 15]
```diff
@@ -4,11 +4,12 @@
  * We use Node's native module system to directly load configuration file.
  * This configuration can (and should) be written as a `.ts` file.
  */
+import fs from '@file-services/node';
```
Comment (AviVahl) [inline packages/scripts/src/application.ts]: move to next import group

File: packages/scripts/src/engine-dashboard/components/features-list.tsx [hunk 16]
```diff
@@ -0,0 +1,116 @@
+import React, { useEffect, useState } from 'react';
+import ReactTable, { Column, DefaultFilterFunction } from 'react-table';
+import 'react-table/react-table.css';
```
Comment (AviVahl) [inline packages/scripts/src/engine-dashboard/components/features-list.tsx]: side-effect imports under all other, if linter allows :)

File: packages/scripts/src/engine-dashboard/components/features-list.tsx [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+import {
+    isListNodeEnvironmtnrsResponse,
+    isPossibleFeaturesAndConfigs,
+    isSuccessResponse,
+    PossibleFeaturesAndConfigs,
+    ServerResponse
+} from '../../server-types';
+import { ServerEnvironmentButton } from './button';
+
+const fetchAllPossibleFeatures = async () => (await fetch('possible-entities')).json();
```
Comment (AviVahl) [inline packages/scripts/src/engine-dashboard/components/features-list.tsx]: use /

File: packages/scripts/src/engine-dashboard/components/features-list.tsx (hunk 16 above)
Comment (RomanYarik) [reply packages/scripts/src/engine-dashboard/components/features-list.tsx]: doesn't allow :)

**Pull request merged**